core/runtime/bindings.rs

```diff
@@ -105,11 +110,47 @@ pub(crate) fn create_external_references(
     }
   }
 
+  for source in sources {
+    references.push(v8::ExternalReference {
+      pointer: source as *const _ as _,
+    })
+  }
+
   references.extend_from_slice(additional_references);
 
-  let refs = v8::ExternalReferences::new(&references);
-  let refs: &'static v8::ExternalReferences = Box::leak(Box::new(refs));
-  refs
+  v8::ExternalReferences::new(&references)
+}
+
+pub(crate) fn externalize_sources(
+  sources: &mut LoadedSources,
+  snapshot_sources: Vec<&'static [u8]>,
+) -> (Box<[v8::OneByteConst]>, Box<[FastString]>) {
+  const INIT_VALUE: MaybeUninit<[usize; 3]> =
+    MaybeUninit::<[usize; 3]>::uninit();
+  let externals =
+    vec![INIT_VALUE; sources.len() + snapshot_sources.len()].into_boxed_slice();
+  let mut original_sources = Vec::with_capacity(sources.len());
+  unsafe {
+    let mut externals: Box<[v8::OneByteConst]> = std::mem::transmute(externals);
+    for (index, source) in sources.into_iter().enumerate() {
+      externals[index] = v8::String::create_external_onebyte_const(
+        std::mem::transmute(source.code.as_bytes()),
+      );
+      let ptr = &externals[index] as *const v8::OneByteConst;
+      let original_source = std::mem::replace(
+        &mut source.code,
+        FastStaticString::from(&*ptr).into(),
+      );
+      original_sources.push(original_source)
+    }
+    let offset = sources.len();
+    for (index, source) in snapshot_sources.into_iter().enumerate() {
+      externals[index + offset] =
+        v8::String::create_external_onebyte_const(source);
+    }
+
+    (externals, original_sources.into_boxed_slice())
+  }
```

Comment:
Can you add comment here and explain what's going on here? I don't get it. Also a `// SAFETY` comment on `unsafe` block is needed I think